ehrapy transition to ehrdata: take over ehrapys data functionalities

eroell opened the pull request:
Fixes #134, fixes #130, fixes #133.

The data specific functionality of ehrapy (modules `ehrapy.io`, `ehrapy.anndata`, `ehrapy.datasets`) should be moved to ehrdata.
These functionalities are closely intertwined, and build on top of each other.

This should be a first larger PR to port and adapt these functions to EHRData in one go, where the dependencies between the modules are visible at one glance and can be accounted for together.

Later PRs can then add more functionality and fix edge cases.
| io ehrapy function                     |   ehrdata function     | Notes | 
|------------------------------------|--------|-------|
| `io.read_csv`   |    `io.read_csv`    | Large simplification    |
| `io.read_h5ad`  |   `io.read_h5ad`     | Large simplification    | 
| No equivalent  |   `io.read_zarr`     | New    | first draft         |
| `io.write`       |    -    | Large simplification, split up to `write_h5ad` and `write_zarr`     | 
| `io.write`  |   `io.write_h5ad`     | Large simplification    | 
| `io.write`  |   `io.write_zarr`     | New    | 


| anndata ehrapy function                     |   ehrdata function     | Notes | 
|------------------------------------|--------|-------|
| `anndata.infer_feature_types` |    `ed.infer_feature_types`    |  slight changes for missing data and string format improvement | 
| `anndata.feature_type_overview`   |    `ed.feature_type_overview`     |  slight changes for missing data and string format improvement   | 
| `anndata.replace_feature_types`   |  `ed.replace_feature_types`      |  almost copy-paste   |   
| no equivalent                      |  `ed.harmonize_missing_values`      | New function for dealing with strings that indicate absence of value    |  
| `anndata.df_to_anndata`           |   `ed.io.from_pandas`        |  Now supports longitudinal dataframe reading, in two accepted input formats.       |
| `anndata.anndata_to_df`           |  `ed.io.to_pandas`      | change what `var_col` does, its use didnt make sense in ehrapy.  Now supports longitudinal dataframe writing to two output formats.          |

| datasets ehrapy function                     |   ehrdata function     | Notes | 
|------------------------------------|--------|-------|
| `datasets.mimic_2`  |   `datasets.mimic_2`     | Returns `EHRData` instead of `AnnData`    | 
| `datasets.mimic_2_preprocessed`  |   `datasets.mimic_2_preprocessed`     | Returns `EHRData` instead of `AnnData`    | 
| `datasets.diabetes_130_raw`  |   `datasets.diabetes_130_raw`     | Returns `EHRData` instead of `AnnData`    | 
| `datasets.diabetes_130_fairlearn`  |   `datasets.diabetes_130_fairlearn`     | Returns `EHRData` instead of `AnnData`    |

eroell commented on src/ehrdata/core/constants.py:
```diff
@@ -1 +1,23 @@
+# --------------
+# These values if encountered as strings are considered to represent missing values in the data
+
+MISSING_VALUES = (
```
What are your opinions on this list?
I think this list is not matured, but a starting point.

This is helpful particularly since in a first draft of the read/write dtype-object arrays, some casting of arrays to strings and back happens.

eroell commented on src/ehrdata/io/zarr.py:
```diff
@@ -0,0 +1,105 @@
+        >>> ed.io.write_zarr("mimic_2.zarr", edata)
+    """
+    filename = Path(filename)  # allow passing strings
+
```
See question at `write_zarr` for discussion of how to write dtype object arrays

eroell commented on src/ehrdata/io/h5ad.py:
```diff
@@ -0,0 +1,104 @@
+                edata.layers[layer] = array.astype(np.float64)
+            except ValueError:
+                edata.layers[layer] = array.astype(str)
+    ad.AnnData(edata).write_h5ad(
```
What is your take on this writing in two steps?
I hoped it leverages as much as possible of `AnnData` utility. There is only the additional `tem` field we need to write somehow.

eroell replied:
So the two sequential write calls without further control are dangerous?

How could a try-finally block help? Ensuring that either `tem` is successfully written too, and otherwise the complete write fails?

Zethson commented on src/ehrdata/tl/pandas/_from_pandas.py:
```diff
@@ -0,0 +1,186 @@
+
+    # Prepare the AnnData object
+    if format == "flat":
+        X = df.to_numpy(copy=True)
```
Another copy?

Zethson commented on src/ehrdata/tl/pandas/_to_pandas.py:
```diff
@@ -0,0 +1,100 @@
+
+
+def to_pandas(
+    edata: EHRData,
```
Asterix? (Not Obelix).

Help, I lost me cheese in the Genfer See

eroell replied:
Yes. I generally love asterix, both to make sure order of arguments is not important if the arguments are not naturally ordered, as well as reading the comics.

Zethson commented on src/ehrdata/tl/pandas/_to_pandas.py:
```diff
@@ -0,0 +1,100 @@
+        format: The format of the output DataFrame. This is relevant for longitudinal data. If "wide", the output dataframe will write a column for each (variable, time) tuple, naming the colun as <variable_name>_t_<tem.index value>. If "long", the output dataframe will be in long format, with columns "observation_id", "variable", "time", and "value".
+
+    Returns:
+        The data object as a pandas DataFrame.
```
Not that helpful

Zethson commented on src/ehrdata/tl/pandas/_to_pandas.py:
```diff
@@ -0,0 +1,100 @@
+    Returns:
+        The data object as a pandas DataFrame.
+
+    Examples:
```
More examples?

Zethson commented on src/ehrdata/tl/pandas/_to_pandas.py:
```diff
@@ -0,0 +1,100 @@
+    var_names = edata.var_names if var_col is None else edata.var[var_col]
+
+    if issparse(X):  # pragma: no cover
+        X = X.toarray()
```
Use fast-array-utils.

IDK whether we want to depend on it here but maybe the non numba version (without accel extra)

Zethson reviewed (changes requested):
Really big PR! Congratulations - many steps forward.

1. Generally, the examples could also show the output. Then they're easier to understand. Like what kinda DataFrame do I get back? What does my EHRData object look like?
2. Let's also check code coverage

eroell commented on CHANGELOG.md:
```diff
@@ -8,6 +8,30 @@ and this project adheres to [Semantic Versioning][].
+
+### Fixed
+
+### Added
```
Will do

eroell commented on docs/api/tools_index.md:
```diff
@@ -10,6 +10,12 @@
     :toctree: io
     :nosignatures:
 
+    tl.infer_feature_types
```
Did not put the pandas in IO, since IO for me meant basically disk. So reading a csv is IO, but a pandas dataframe transformation is not. I don't have strong opinions here, just writing thoughts in words what you probably inferred from code already.

Regarding tools, looking for a meaningful module name, thought this is actually a nice wording for these kinds of functions.
Interested to hear other opinions than what's drafted

eroell commented on src/ehrdata/dt/dataloader.py:
```diff
@@ -16,13 +16,16 @@
+RAW_FORMATS_LIST = list(get_args(RAW_FORMATS))
 
 
 def download(
```
Yes agree, will do

eroell commented on src/ehrdata/dt/dataloader.py:
```diff
@@ -16,13 +16,16 @@
 
 COMPRESSION_FORMATS = Literal["tar.gz", "gztar", "zip", "tar", "gz", "bz", "xz"]
 COMPRESSION_FORMATS_LIST = list(get_args(COMPRESSION_FORMATS))
+RAW_FORMATS = Literal["csv", "txt", "parquet", "h5ad"]
```
because I forgot

eroell commented on src/ehrdata/dt/datasets.py:
```diff
@@ -527,3 +528,136 @@ def physionet2012(
+
+    Examples:
+        >>> import ehrdata as ed
+        >>> adata = ep.dt.mimic_2(encoded=True)
```
Yep

flying-sheep commented on src/ehrdata/tl/pandas/_from_pandas.py:
```diff
@@ -0,0 +1,186 @@
+
+
+def from_pandas(
+    df: pd.DataFrame,
```
btw: it's "asterisk". Just in case this isn't a continuation of the joke above.

Zethson commented on src/ehrdata/core/ehrdata.py:
```diff
@@ -29,7 +29,7 @@ class EHRData(AnnData):
        :class: dark-light
 
-    Extends :class:`~anndata.AnnData` to further support regular and irregular time-series data.
+    Extends  to further support regular and irregular time-series data.
```
What happened here?

Zethson commented on src/ehrdata/dt/dataloader.py:
```diff
@@ -16,11 +16,11 @@
 
 
-def download(
+def _download(
```
I remember my suggestion but it's probably cleaner if the whole dataloader module is private? Users should never interact with it, right?

Zethson reviewed (commented):
Github told me that a review was requested again. If that's the case, there's still a few unaddressed comments of mine. If you don't want to tackle them with his PR, please comment and/or resolve them, please.

eroell commented:
> Github told me that a review was requested again. If that's the case, there's still a few unaddressed comments of mine. If you don't want to tackle them with his PR, please comment and/or resolve them, please.

That was not intended, had one commit yesterday, but am not yet done with addressing comments.

eroell commented:
Dear both,

I’d be very interested in your comments in a second review round.

@flying_sheep, I’d particularly be interested for your opinion and comments on 4 functions:
- `read_h5ad` and `write_h5ad` in `src/ehrdata/io/h5ad.py`
- `read_zarr` and `write_zarr` in src/ehrdata/io/zarr.py`

eroell commented on src/ehrdata/io/h5ad.py:
```diff
@@ -0,0 +1,104 @@
+    with h5py.File(filename, "r") as f:
+        dictionary_for_init = {k: ad.io.read_elem(f[k]) for k, v in dict(f).items() if not k.startswith("raw.")}
+
+    # If X, layers is str; convert to object dtype. First, try to cast each column to float64 in pandas.
```
I intended to make the entire casting behaviour is much more explicit now.

The casting is now an option, by default set to `True`.

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+
+
+def _detect_feature_type(col: pd.Series) -> tuple[Literal["date", "categorical", "numeric"], bool]:
+    """Detect the feature type of a column in a pandas DataFrame.
```
```suggestion
    """Detect the feature type of a :class:`~pandas.Series`.
```

Double typing it is dumb but a column of a DF is a Series. You can add more details on the detected types and how it works in the docstring if you so desire.

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+    """Detect the feature type of a column in a pandas DataFrame.
+
+    Args:
+        col: The column to detect the feature type for.
```
Yes, but it's any Series

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+    recommended to check the inferred types.
+
+    Args:
+        edata: Data object.
```
I wonder whether we can have something more descriptive than that. I like the AnnData version.

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+    X = edata.X if layer is None else edata.layers[layer]
+
+    if issparse(X):
+        X = X.toarray()
```
fast-array-utils

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+    if verbose:
+        logger.warning(
+            f"{'Features' if len(uncertain_features) > 1 else 'Feature'} {str(uncertain_features)[1:-1]} {'were' if len(uncertain_features) > 1 else 'was'} detected as categorical features stored numerically."
+            f"Please verify and correct using `ed.replace_feature_types` if necessary."
```
This and the info message just below have different sentence structure for no good reason. It's also a tiny bit verbose but that's a nit.

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+        raise ValueError(err_msg)
+
+
+# TODO: this function is a different flavor of inferring feature types. We should decide on a single implementation in the future.
```
Probably the 5th time I'm writing that comment. I'll now just assume that you'll solve that later with the ehrapy PR. Trust me, I'll annoy you about this haha.

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+    Args:
+        edata: Data object.
+        features: The features to correct.
+        corrected_type: The corrected feature type. One of `'date'`, `'categorical'`, or `'numeric'`.
```
If that's true, this should be a literal and not a str

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+
+    This function will replace strings that are considered to represent missing values with `np.nan`.
+
+
```
```suggestion
```

Zethson commented on src/ehrdata/_feature_types.py:
```diff
@@ -0,0 +1,326 @@
+    """
+    if copy:
+        edata = edata.copy()
+    X = edata.X if layer is None else edata.layers[layer]
```
We'll need an internal function for this as soon as Phil's PR is merged. Let's not forget about this. You can resolve this comment.

Zethson commented on src/ehrdata/dt/datasets.py:
```diff
@@ -23,6 +24,8 @@
 from scipy.sparse import csr_matrix
 from sparse import COO
 
+DEFAULT_DATA_PATH = Path("ehrapy_data")
```
I wonder whether that should be in a `constants.py` module

Zethson commented on src/ehrdata/dt/datasets.py:
```diff
@@ -524,6 +519,131 @@ def physionet2012(
+    Args:
+        columns_obs_only: Columns to include only in obs and not X.
+
+
```
```suggestion
```

Zethson commented on src/ehrdata/dt/datasets.py:
```diff
@@ -524,6 +519,131 @@ def physionet2012(
+
+    Examples:
+        >>> import ehrdata as ed
+        >>> edata = ep.dt.mimic_2()
```
```suggestion
        >>> edata = ed.dt.mimic_2()
```

Zethson commented on src/ehrdata/dt/datasets.py:
```diff
@@ -524,6 +519,131 @@ def physionet2012(
+
+    More details on the original dataset: https://physionet.org/content/mimic2-iaccd/1.0/.
+
+
```
```suggestion
```

Zethson commented on src/ehrdata/dt/datasets.py:
```diff
@@ -524,6 +519,131 @@ def physionet2012(
+    Args:
+        columns_obs_only: Columns to include in `obs` only and not `X`.
+
+
```
```suggestion
```

Zethson commented on src/ehrdata/dt/datasets.py:
```diff
@@ -524,6 +519,131 @@ def physionet2012(
+        >>> import ehrdata as ed
+        >>> edata = ed.dt.diabetes_130_fairlearn()
+
+    References:
```
All of these should be cited properly. We have a references.bib. Put stuff there and cite it

Zethson commented on src/ehrdata/io/_array_casting.py:
```diff
@@ -0,0 +1,47 @@
+from lamin_utils import logger
+
+
+def _cast_variables_to_float(edata: "EHRData") -> None:
```
```suggestion
def _cast_variables_to_float(edata: EHRData) -> None:
```

Why?

eroell replied:
I get an error with that, and I have no idea why, its just this file. My only remaining take is to turn off and on my computer

Zethson commented on src/ehrdata/io/_array_casting.py:
```diff
@@ -0,0 +1,47 @@
+                    edata.layers[key][:, column] = edata.layers[key][:, column].astype(np.float64)
+
+
+def _cast_arrays_dtype_to_float_or_str_if_nonnumeric_object(edata: "EHRData") -> "EHRData":
```
```suggestion
def _cast_arrays_dtype_to_float_or_str_if_nonnumeric_object(edata: EHRData) -> EHRData:
```

Zethson commented on src/ehrdata/io/csv.py:
```diff
@@ -0,0 +1,57 @@
+    | None = None,
+    **kwargs,
+) -> EHRData:
+    """Reads a csv file.
```
```suggestion
    """Reads a csv file.
```

Read a comma-separated values (csv) file into DataFrame.

is what the pandas docs have: https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

So let's write it here accordingly to make it obvious that we get an EHRData object. Yes, the return types says it but this function has lots of parameters and one might need to scroll

Zethson commented on src/ehrdata/io/csv.py:
```diff
@@ -0,0 +1,57 @@
+    """Reads a csv file.
+
+    This function reads a csv file, and creates an :class:`ehrdata.EHRData` object.
+    It first reads the csv file using :func:`pandas.read_csv`, and then passes the resulting :class:`~pandas.DataFrame` to :func:`ehrdata.io.from_pandas`.
```
Implementation detail. Not relevant for users.

eroell replied:
thought this is very explicit about how reading a csv is done, and hints the option of doing it yourself in two steps

Zethson commented on src/ehrdata/io/csv.py:
```diff
@@ -0,0 +1,57 @@
+) -> EHRData:
+    """Reads a csv file.
+
+    This function reads a csv file, and creates an :class:`ehrdata.EHRData` object.
```
That's what the first line of the docstring already said. (or should say)
```suggestion
```

Zethson commented on src/ehrdata/io/csv.py:
```diff
@@ -0,0 +1,57 @@
+        sep: Separator in the file. Delegates to :func:`pandas.read_csv`.
+        index_column: If specified, this column of the csv file will be used for the `.obs` dataframe. Delegates to :func:`~ehrdata.io.from_pandas`.
+        columns_obs_only: These columns will be added to the `.obs` dataframe only. Delegates to :func:`~ehrdata.io.from_pandas`.
+        format: The format of the input dataframe. If the data is not longitudinal, choose `format="flat"`. If the data is longitudinal in the long format, choose `format="long"`. If the data is longitudinal in a wide format, choose `format="wide"`. Delegates to :func:`~ehrdata.io.from_pandas`.
```
Long ass line. Consider splitting into multiple. Also applies to other lines here.

Zethson commented on src/ehrdata/io/csv.py:
```diff
@@ -0,0 +1,57 @@
+
+    Examples:
+        >>> import ehrdata as ed
+        >>> edata = ed.io.read_csv("myfile.csv")
```
Even cooler if this was a standalone example where we get a CSV file from somewhere but let's not make it complex. Please resolve but I wanted to write this nevertheless.

Zethson reviewed (approved):
Coolio! We're getting there.

1. Please always write sentences in their own lines in markdown. RE your notebooks
2. You can and should also use intersphinx in notebooks :)
3. Use proper citations, please.
4. Consider searching for `ep.dt` to ensure that these are all gone

eroell commented:
The readthedocs build is not shown in the checks here, but I can see on the readthedocs project that it is built and succeeds
